src/IB/IBImplicitHierarchyIntegrator.cpp

```diff
@@ -304,160 +307,63 @@ IBImplicitHierarchyIntegrator::integrateHierarchy(const double current_time, con
-        d_ib_implicit_ops->createSolutionVec(&DX_nm1);
-        d_ib_implicit_ops->createSolutionVec(&DX_n);
-        d_ib_implicit_ops->createSolutionVec(&DX_np1);
-        d_ib_implicit_ops->createSolutionVec(&DDX_n);
-        d_ib_implicit_ops->createSolutionVec(&DGX_n);
-
-        // Initialize X_{n-1}:
-        d_ib_implicit_ops->getUpdatedPosition(X_nm1);
-
-        double D0;
-        ierr = VecNorm(X_nm1, NORM_2, &D0);
-        IBTK_CHKERRQ(ierr);
-
-        int k = 0, n_solves = 0;
-        double tol = 1.0e-5;
-        int max_its = 100;
-        bool converged = false;
-
-        // Initialize X_{n} = G(X_{n-1}):
-        iterateSolution(X_nm1);
-        d_ib_implicit_ops->getUpdatedPosition(GX_nm1);
-        ++n_solves;
-        ierr = VecCopy(GX_nm1, X_n);
-        IBTK_CHKERRQ(ierr);
-
-        // Initialize DX_{n-1} := G(X_{n-1}) - X_{n-1}:
-        ierr = VecWAXPY(DX_nm1, -1.0, X_nm1, GX_nm1);
-        IBTK_CHKERRQ(ierr);
-
-        for (k = 0; k < max_its && !converged; ++k)
-        {
-            // G(X_{n}):
-            iterateSolution(X_n);
-            d_ib_implicit_ops->getUpdatedPosition(GX_n);
-            ++n_solves;
-
-            // DX_{n} := G(X_{n}) - X_{n}:
-            ierr = VecWAXPY(DX_n, -1.0, X_n, GX_n);
-            IBTK_CHKERRQ(ierr);
-
-            // DGX_{n} := G(X_{n}) - G(X_{n-1}):
-            ierr = VecWAXPY(DGX_n, -1.0, GX_nm1, GX_n);
-            IBTK_CHKERRQ(ierr);
-
-            // DDX_{n} := DX_{n} - DX_{n-1}:
-            ierr = VecWAXPY(DDX_n, -1.0, DX_nm1, DX_n);
-            IBTK_CHKERRQ(ierr);
-
-            // X_{n+1} := G(X_{n}) - Q (G(X_{n}) - G(X_{n-1}))
-            //          = G(X_{n}) - Q DGX_{n}
-            //
-            // with Q = dot(DX_{n} - DX_{n-1}, DX_{n})/||DX_{n} - DX_{n-1}||^2
-            //        = dot(DDX_{n}, DX_{n})/||DDX_{n}||^2
-
-            double Q1;
-            ierr = VecDot(DDX_n, DX_n, &Q1);
-            IBTK_CHKERRQ(ierr);
-            double Q2;
-            ierr = VecDot(DDX_n, DDX_n, &Q2);
-            IBTK_CHKERRQ(ierr);
+            else if (d_implicit_algorithm == "IRONS-TUCK")
+            {
+                // X_{n+1} := omega_{n} G(X_n) + (1.0-omega_{n}) X_n
+                //
+                // with omega_{n} = -omega_{n-1} dot(DX_{n-1}, DX_{n} - DX_{n-1})/||DX_{n} - DX_{n-1}||^2
+                //                = -omega_{n-1} dot(DX_{n-1}, DDX_{n})/||DDX_{n}||^2
+                double Q1;
+                ierr = VecDot(DX_nm1, DDX_n, &Q1);
+                IBTK_CHKERRQ(ierr);
+                double Q2;
+                ierr = VecDot(DDX_n, DDX_n, &Q2);
+                IBTK_CHKERRQ(ierr);
 
-            const double Q = Q1 / Q2;
-            ierr = VecWAXPY(X_np1, -Q, DGX_n, GX_n);
-            IBTK_CHKERRQ(ierr);
+                d_omega_n = -omega_nm1 * Q1 / Q2;
+                ierr = VecAXPBYPCZ(X_np1, d_omega_n, 1.0 - d_omega_n, 0.0, GX_n, X_n);
+                IBTK_CHKERRQ(ierr);
+                omega_nm1 = d_omega_n;
```

Comment:
Maybe for k==0 we only need to use the one from previous time step like the following code.

  ```
  else if (d_implicit_algorithm == "IRONS-TUCK")
            {
                if(k>0)
                {
                // X_{n+1} := omega_{n} G(X_n) + (1.0-omega_{n}) X_n
                //
                // with omega_{n} = -omega_{n-1} dot(DX_{n-1}, DX_{n} - DX_{n-1})/||DX_{n} - DX_{n-1}||^2
                //                = -omega_{n-1} dot(DX_{n-1}, DDX_{n})/||DDX_{n}||^2
                double Q1;
                ierr = VecDot(DX_nm1, DDX_n, &Q1);
                IBTK_CHKERRQ(ierr);
                double Q2;
                ierr = VecDot(DDX_n, DDX_n, &Q2);
                IBTK_CHKERRQ(ierr);
            const double Q = Q1 / Q2;
            ierr = VecWAXPY(X_np1, -Q, DGX_n, GX_n);
            IBTK_CHKERRQ(ierr);
                d_omega_n = -omega_nm1 * Q1 / Q2;
                ierr = VecAXPBYPCZ(X_np1, d_omega_n, 1.0 - d_omega_n, 0.0, GX_n, X_n);
                IBTK_CHKERRQ(ierr);
                omega_nm1 = d_omega_n;
                }else
               {
                d_omega_n=omega_nm1; 
                }
```

Reply:
It is pretty cheap to compute the updated value of `omega`, so I think we might as well do it unless this improves convergence.

Reply:
I think we should put the following code into integrateHiearchy
```
    Vec X_old_time_step;
   d_ib_implicit_ops->createSolutionVec(X_old_time_step);
   d_ib_implicit_ops->getUpdatedPosition(X_old_time_step);
    if (d_use_structure_predictor)
    {
        if (d_enable_logging)
            plog << d_object_name << "::preprocessIntegrateHierarchy(): performing Lagrangian forward Euler step\n";
        d_ib_implicit_ops->forwardEulerStep(current_time, new_time);
    }

```
Also we need to delete the predictor in the 
```

IBImplicitHierarchyIntegrator::preprocessIntegrateHierarchy(const double current_time,
                                                            const double new_time,
                                                            const int num_cycles)

```

Reply:
This is a little opaque in the interface, but the IB implementation keeps track of `X^{n}` and `X^{n+1}` internally and always maintains `X^{n+1/2} = (X^{n+1}+X^{n})/2` for the current values of `X^{n}` and `X^{n+1}`.

`setUpdatedPosition(X_updated)` will set the current approximation to `X^{n+1}` used in the IB implementation to the provided position vector `X_updated`: `X^{n+1} := X_updated`.

`getUpdatedPosition(X_updated)` does the opposite; it will get : `X_updated := X^{n+1}`.

`iterateSolution(X_updated)` will do a basic momentum equation solve using the provided value of `X_updated` as the approximation to `X^{n+1}` for the purposes of evaluating the structural forces, and for spreading and interpolating forces. Then, the very last thing that `iterateSolution` does is to use a discrete approximation to `dX/dt = U(X,t)` to update the internal approximation to `X^{n+1}`.

So in each call to `iterateSolution()`, there is an initial approximation to `X^{n+1}` and a final approximation to `X^{n+1}`. We declare convergence if the difference between these configurations is relatively small.

Reply:
Note also that the internal vector storing `X^{n}` in the IB implementation is not reset until `IBImplicitHierarchyIntegrator::postprocessIntegrateHierarchy()`.

Reply:
That's right! Storing X^{n} until the subiteration converges is what we want.
Thank you for clarifying this, I am more familiar with the code now.